src/mds/Server.cc

```diff
@@ -6290,6 +6298,149 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
 }
 
+void Server::handle_client_getvxattr(MDRequestRef& mdr)
+{
+  const auto& req = mdr->client_request;
+  string name{req->get_path2()};
+
+  // is a ceph virtual xattr?
+  if (!is_ceph_vxattr(name)) {
+    respond_to_request(mdr, -CEPHFS_EINVAL);
+    return;
+  }
+
+  CInode *cur = rdlock_path_pin_ref(mdr, true, false);
+  if (!cur) {
+    return;
+  }
+
+  if (!cur->is_dir()) {
+    respond_to_request(mdr, -CEPHFS_ENOTDIR);
+    return;
+  }
+
+  std::stringstream ss;
+  int r = 0;
+  ceph::bufferlist bl;
+  // handle these vxattrs
+  if (name.substr(0, 15) == "ceph.dir.layout"sv) {
+    std::string layout_field;
+
+    struct layout_xattr_info_t {
+      const file_layout_t  layout;
+      const std::string    inheritance;
+
+      layout_xattr_info_t(const file_layout_t& l, const std::string& inh)
+        : layout(l), inheritance(inh) { }
+    };
+
+    auto is_default_layout = [](const file_layout_t& a) -> bool {
+      auto default_layout = file_layout_t::get_default();
+      return (a.stripe_unit  == default_layout.stripe_unit and
+	      a.stripe_count == default_layout.stripe_count and
+	      a.object_size  == default_layout.object_size);
+    };
+    auto get_inherited_layout = [&](CInode *cur) -> layout_xattr_info_t {
+      auto orig_in = cur;
+
+      while (cur) {
+        if (cur->get_projected_inode()->has_layout()) {
+	  auto& curr_layout = cur->get_projected_inode()->layout;
+          if (cur == orig_in) {
+	    if (is_default_layout(curr_layout)) {
+	      return {mdcache->default_file_layout, "default"s};
+	    } else {
+	      // we've found a new layout at this inode
+	      return {cur->get_projected_inode()->layout, "new"s};
+	    }
+          } else {
+	    if (is_default_layout(curr_layout)) {
+	      return {mdcache->default_file_layout, "default"s};
+	    } else {
+	      return {cur->get_projected_inode()->layout, "inherited"s};
+	    }
+          }
```

Comment:
can't this be simplified as
```
if (is_default_layout(curr_layout)) {
  return {mdcache->default_file_layout, "default"s};
}

if (cur == orig_in) {
  // we've found a new layout at this inode
  return {cur->get_projected_inode()->layout, "new"s};
} else {
   return {cur->get_projected_inode()->layout, "inherited"s};
}
```

Reply:
thanks for the suggestion
updated code as per your suggestion